### ISSUE TITLE
Using css modules like npm package without import

### ISSUE THREAD

**Issue by abraztsov** [css-modules/css-modules#223]
Can I use my `react` component (with `css-modules`) as `npm` package without importing styles ?

### Example

```
import lib from 'my-react-npm-package';
import 'my-react-npm-package/style-that-i-dont-want-to-import.css'
```

**Comment by joeybaker** [comment]
It all depends on your build setup. If you've got things setup to transpile your `node_modules` including a step that runs `import`s through css modules, then there's no need to import the css file on your own. It's just a matter of importing the JS.

**Comment by abraztsov** [comment]
@joeybaker I'm considering a situation when a person that `import` my library doesn't use `webpack` and `css-loader`

**Comment by sullenor** [comment]
In case you have an `index.js` and `index.css` in your package folder, you can not clearly tell what actually the user wants to get from the `import lib from 'my-react-npm-package'`. So it's tricky and depends on the bundling system.

Since the nodejs resolving algorithm checks the `package.json` first (if you provide a folder as a path), the `main` field can be used here to point resolver what the file should be loaded. But, I'm not sure whether this behavior is supported for the css files. 

> I don't want to write in the README that you need to put a css loader , styles loader and a bunch of other loaders just for my library to be able to use styles.

You will have to do that cause you can't load a plain CSS file without any configuration.

**Comment by abraztsov** [comment]
@sullenor 

> UP: At least you can mention, that you use css modules and point any popular guide with configuration

This is the problem that I want to solve. I'm talking about this above.
As a developer I just want to `import` this `package` and use it. Easy. Clear. Good way.

A large percentage of people don't want to read text like this `For using my package you need to read my guide and then this guide and this and little of this`. The simpler the better

I think you and a lot of developer understand what I'm talking about.

The main question is still here. Is it real to my package `auto import and use` styles without additional configuration ?

**Comment by sullenor** [comment]
> A large percentage of people don't want to read text like this For using my package you need to read my guide and then this guide and this and little of this. The simpler the better.

This works till the people know how to use your project and resolve its dependencies. For example, to use `import` they know, that they need to setup babel or webpack (since it's not implemented yet with any environment). Same thing for `require` (in browser env) and CSS. 
So, I doubt that it is possible to find any suitable solution for your goal.

But, if we take a webpack as example, then the issue can be easily solved. If you specify `main` field in the `package.json` of your lib, like:

```json
{
  "name": "lib",
  "main": "my-lib.css",
}
```

then webpack will handle it properly with basic `css-loader` configuration, like

```js
{
  loader: 'css-loader',
  options: {
    modules: true,
  },
}
```

So the `import 'lib'` will automatically resolve into `import 'lib/my-lib.css'`.

**Comment by joeybaker** [comment]
@abraztsov The short answer to your question is "no". CSS modules expects there to be some build step that is capable of understanding what importing a css file means. That's not something that JS does by default, so there's way around this except with tooling.

The good news is that this tooling is pretty common now, to the point where I'd say that all modern projects expect something capable of this to be available. 

The only answer that I know of that allows you to just allow a user to require your lib and get CSS at the same time is inline styles. That's an option if you absolutely don't want to require a build step. However, they come that approach comes with it's own suite of problems.

TL;DR: There is no perfect answer right now.

**Comment by abraztsov** [comment]
> For example, to use import they should know, that they need to setup babel or webpack
> The good news is that this tooling is pretty common now, to the point where I'd say that all modern projects expect something capable of this to be available.

Agree.

As a result - it's impossible right now.

**Comment by bryzaguy** [comment]
Why not bundle style loader configured as a singleton, without hmr, and the webpack library target as commonjs? This way there are no build requirements for consumers. This has worked really well for me. Is there anything wrong with that approach?

**Comment by pqt** [comment]
@bryzaguy do you have an example of this configuration in action?
